### PR TITLE
Try to fix link issues with qt6 resources

### PR DESCRIPTION
**Pull request by sloriot**

### CODE REVIEW

**Review by lrineau — APPROVED** [review]

**Comment by lrineau** [comment]
I think I have eventually found the culprit lines: see the fix in 811f78c6a1d9d6c9cb0585062de0f07b45a558e7.

## Before

Before that commit, the following command

```
rm -rf /tmp/build || true; cmake -DCGAL_TEST_SUITE=ON -S ~/Git/cgal-master -B /tmp/build -DWITH_examples=ON -DCGAL_ENABLE_TESTING=ON && ctest --test-dir /tmp/build -j12 -L Linear_cell_complex_Examples --stop-on-failure --output-on-failure
```

was failing, and the first line of the output of CTest were:

```
Internal ctest changing into directory: /tmp/build
Test project /tmp/build
      Start   2: check build system
      Start 768: copy source_dir of Linear_cell_complex_Examples
 1/29 Test #768: copy source_dir of Linear_cell_complex_Examples ...............   Passed    0.01 sec
 2/29 Test   #2: check build system ............................................   Passed    0.16 sec
      Start 766: compilation of  gmap_linear_cell_complex_3
      Start 770: compilation of  linear_cell_complex_3
      Start 772: compilation of  linear_cell_complex_3_attributes_management
      Start 774: compilation of  linear_cell_complex_3_incremental_builder
      Start 776: compilation of  linear_cell_complex_3_insert
      Start 778: compilation of  linear_cell_complex_3_operations
      Start 780: compilation of  linear_cell_complex_3_with_colored_vertices
      Start 782: compilation of  linear_cell_complex_3_with_mypoint
      Start 784: compilation of  linear_cell_complex_4
      Start 786: compilation of  plane_graph_to_lcc_2
      Start 788: compilation of  voronoi_2
      Start 790: compilation of  voronoi_3
 3/29 Test #780: compilation of  linear_cell_complex_3_with_colored_vertices ...***Failed   24.13 sec
[  0%] Automatic MOC for target CGAL_Qt6_moc_and_resources
[  0%] Built target CGAL_Qt6_moc_and_resources_autogen
[  0%] Building CXX object examples/Arrangement_on_surface_2/CMakeFiles/CGAL_Qt6_moc_and_resources.dir/CGAL_Qt6_moc_and_resources_autogen/mocs_compilation.cpp.o
[100%] Linking CXX static library libCGAL_Qt6_moc_and_resources.a
/usr/bin/ranlib: unable to copy file 'libCGAL_Qt6_moc_and_resources.a'; reason: No such file or directory
gmake[3]: *** [examples/Arrangement_on_surface_2/CMakeFiles/CGAL_Qt6_moc_and_resources.dir/build.make:203: examples/Arrangement_on_surface_2/libCGAL_Qt6_moc_and_resources.a] Error 1
gmake[3]: *** Deleting file 'examples/Arrangement_on_surface_2/libCGAL_Qt6_moc_and_resources.a'
gmake[2]: *** [CMakeFiles/Makefile2:6414: examples/Arrangement_on_surface_2/CMakeFiles/CGAL_Qt6_moc_and_resources.dir/all] Error 2
gmake[1]: *** [CMakeFiles/Makefile2:14065: examples/Linear_cell_complex/CMakeFiles/linear_cell_complex_3_with_colored_vertices.dir/rule] Error 2
gmake: *** [Makefile:4690: linear_cell_complex_3_with_colored_vertices] Error 2
```

## After

After that commit 811f78c6a1d9d6c9cb0585062de0f07b45a558e7, the same command succeeds, and the first lines of CTest are:

```
Internal ctest changing into directory: /tmp/build
Test project /tmp/build
      Start   2: check build system
      Start 769: copy source_dir of Linear_cell_complex_Examples
 1/30 Test #769: copy source_dir of Linear_cell_complex_Examples ...............   Passed    0.01 sec
 2/30 Test   #2: check build system ............................................   Passed    0.16 sec
      Start 109: compilation of  CGAL_Qt6_moc_and_resources
 3/30 Test #109: compilation of  CGAL_Qt6_moc_and_resources ....................   Passed    6.45 sec
      Start 767: compilation of  gmap_linear_cell_complex_3
      Start 771: compilation of  linear_cell_complex_3
      Start 773: compilation of  linear_cell_complex_3_attributes_management
      Start 775: compilation of  linear_cell_complex_3_incremental_builder
      Start 777: compilation of  linear_cell_complex_3_insert
      Start 779: compilation of  linear_cell_complex_3_operations
      Start 781: compilation of  linear_cell_complex_3_with_colored_vertices
      Start 783: compilation of  linear_cell_complex_3_with_mypoint
      Start 785: compilation of  linear_cell_complex_4
      Start 787: compilation of  plane_graph_to_lcc_2
      Start 789: compilation of  voronoi_2
      Start 791: compilation of  voronoi_3
```

The diff is those two lines, that appear very early in the CTest results:

```
      Start 109: compilation of  CGAL_Qt6_moc_and_resources
 3/30 Test #109: compilation of  CGAL_Qt6_moc_and_resources ....................   Passed    6.45 sec
```

The issue before 811f78c6a1d9d6c9cb0585062de0f07b45a558e7 was that the tests added by `cgal_add_test(${exe_name})` were added before the target `${exe_name}` was linked with CGAL imported target (and in particular `CGAL::CGAL_Qt6`).

Now that I have managed to reproduce the bug, and fix it locally, I am now quite confident that we have found the eventual fix.

I still do not understand why it seems the bug was only detected on ArchLinux.

**Comment by sloriot** [comment]
Still not sufficient, see for example [here](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-6.0-Ic-124/Surface_mesh_Examples/TestReport_gimeno_ArchLinux-clang-CXX17-Release.gz)

**Comment by lrineau** [comment]
There is still an issue: all plugins depends on the `Polyhedron_3` target, and that target is built several times in parallel. That can lead to another race-condition, like

https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-6.0-Ic-147/Polyhedron_Demo/TestReport_gimeno_Debian-testing-clang-main.gz

The simplest fix is to remove the dependency to `Polyehdron_3`. @sloriot what do you think?

**Comment by sloriot** [comment]
> There is still an issue: all plugins depends on the `Polyhedron_3` target, and that target is built several times in parallel. That can lead to another race-condition, like
> 
> https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-6.0-Ic-147/Polyhedron_Demo/TestReport_gimeno_Debian-testing-clang-main.gz
> 
> The simplest fix is to remove the dependency to `Polyehdron_3`. @sloriot what do you think?

Looks good to me

File: Polyhedron/demo/Polyhedron/polyhedron_demo_macros.cmake [hunk 1]
```diff
@@ -61,6 +61,9 @@ include(${CGAL_MODULES_DIR}/CGAL_add_test.cmake)
     endif()
     # Link with CGAL
     target_link_libraries( ${plugin_name} PUBLIC CGAL::CGAL )
+    if(CGAL_TEST_SUITE AND TARGET Polyhedron_3)
```
Comment (sloriot) [inline Polyhedron/demo/Polyhedron/polyhedron_demo_macros.cmake:64]: if not?

**Comment by lrineau** [comment]
It seems this time this PR is working correctly, now.

At least in last week testsuite [Test Results of CGAL-6.0-Ic-147, #Polyhedron_Demo](https://cgal.geometryfactory.com/CGAL/testsuite/results-6.0-Ic-147.shtml#Polyhedron_Demo) it seems fine. I have grepped for `Linking.*demo_framework` in all the testresults and it was present only once, in `Polyhedron_Demo`.

```
cgal /srv/CGAL/www/testsuite $ zgrep 'Linking.*demo_framework' CGAL-6.0-Ic-147/*/*ArchLinux-clang*gz   
CGAL-6.0-Ic-147/Polyhedron_Demo/TestReport_gimeno_ArchLinux-clang-CXX17-Release.gz:Linking CXX shared library libdemo_framework.so
CGAL-6.0-Ic-147/Polyhedron_Demo/TestReport_gimeno_ArchLinux-clang.gz:Linking CXX shared library libdemo_framework.so
CGAL-6.0-Ic-147/Polyhedron_Demo/TestReport_gimeno_ArchLinux-clang-Release.gz:Linking CXX shared library libdemo_framework.so
CGAL-6.0-Ic-147/Polyhedron_Demo/TestReport_lrineau_ArchLinux-clang-CXX20-Release.gz:Linking CXX shared library libdemo_framework.so
zsh: exit 1     zgrep 'Linking.*demo_framework' CGAL-6.0-Ic-147/*/*ArchLinux-clang*gz
zsh: exit 1
```

**Pull request merged**